ENH: add Feather IO support

jorisvandenbossche opened the pull request:
@brendan-ward with your hard work from https://github.com/geopandas/geopandas/pull/1180, adding additional Feather support is now actually a breeze ;) 
(the diff is even a bit larger because git doesn't correctly see how the original function was split).

The existing `_to_parquet` is renamed to `_geopandas_to_arrow` with all logic up to having the pyarrow table, and then `_to_parquet` and `_to_feather` have their (little) custom logic to write that table to parquet or feather file, respectively. 
And a similar pattern for the reader.

I think most of the tests can also be parameterized. I already did this for one (the basic roundtrip one).

And maybe can rename `parquet.py` to `arrow.py`

Closes #1409

brendan-ward commented on geopandas/io/parquet.py:
```diff
@@ -181,6 +181,45 @@ def _validate_metadata(metadata):
+    from pyarrow import Table
+
+    warnings.warn(
+        "this is an initial implementation of Parquet file support and "
```
"Parquet" doesn't fit here, and a more generic Arrow-oriented message may not be as obvious to users (since conversion to Arrow table is an implementation detail).  

The warning could be moved to the `_to_parquet` and `_to_feather` level to keep the warning specific to the output format.

jorisvandenbossche replied:
For now I just replaced "Parquet" in the warning an error messages with "Parquet/Feather". 
In principle we could template all those messages to fill in one of the two, as appropriate.

brendan-ward commented on geopandas/io/parquet.py:
```diff
@@ -181,6 +181,45 @@ def _validate_metadata(metadata):
+    metadata = table.schema.metadata
+    metadata.update({b"geo": _encode_metadata(geo_metadata)})
+
+    table = table.replace_schema_metadata(metadata)
```
This could just be returned:

```
return table.replace_schema_metadata(metadata)
```

brendan-ward commented on geopandas/geodataframe.py:
```diff
@@ -609,7 +609,44 @@ def to_parquet(self, filename, compression="snappy", index=None, **kwargs):
+        Parameters
+        ----------
+        path : str, path object
+        compression : {'snappy', 'gzip', 'brotli', None}, default 'snappy'
```
The compression options don't match the spec for Feather, if I'm reading it correctly:
https://arrow.apache.org/docs/python/generated/pyarrow.feather.write_feather.html

```
'zstd', 'lz4', 'uncompressed'
```

It looks like we specifically have to map `None` here to `uncompressed` for output, as the default behavior for `write_feather` is `lz4` if it is available.

jorisvandenbossche replied:
I would leave the default to match with pyarrow.feather (the default of None means: use lz4 if available, otherwise uncompressed). Updated the docstring for that.

brendan-ward replied:
It looks like the docstring got updated for `_to_feather` but not `to_feather` also.

Having the compression defaults between Parquet / Feather bugs me a little bit from an API consistency standpoint where `compression=None` semantically reads as no compression, rather than use the default of the underlying lib.

But!  The performance for Feather + LZ4 looks _so_ much better that it's hard to argue against that as a sensible default.

Would it be too ugly to add `"uncompressed"` as an option for Parquet, where `compression=None` means "use the default" (which happens to be uncompressed) and `compression="uncompressed"` means explicitly do not compress it?

brendan-ward replied:
@jorisvandenbossche it might be github hiding things from me, but it looks like this docstring for `to_feather` at the top level needs to be updated to match `_to_feather`.

jorisvandenbossche replied:
> docstring for to_feather at the top level needs to be updated to match _to_feather.

Yes, good catch, fixed.

Regarding the inconsistency about None vs "uncompressed", that's a good point. But I would rather like to bring this up with pyarrow to fix it there.

Of course we could easily add "uncompressed" to `to_parquet` here as well (but it would also be inconsistent with pandas, unless we add it there as well ..)

brendan-ward replied:
👍 on taking this up in `pyarrow` instead.

brendan-ward commented on geopandas/io/parquet.py:
```diff
@@ -215,88 +254,64 @@ def _to_parquet(df, path, compression="snappy", index=None, **kwargs):
 
-    kwargs["use_pandas_metadata"] = True
-    table = parquet.read_table(path, columns=columns, **kwargs)
+    if pyarrow.__version__ < "0.17":
```
Sorry - this was my fault.  I opted to go simplistic in porting `import_optional_dependency` from `pandas`; we now need version support there.  Crossref: https://github.com/pandas-dev/pandas/blob/master/pandas/compat/_optional.py#L47-L111

jorisvandenbossche replied:
No, no, I actually asked you to keep it simple ;) We should only port what we needed, and in the previoius PR, this was not yet needed. 

Now, the version matching can also not be copied from pandas, as it works a bit differently there (also, we are importing a submodule which has not a version itself .. which makes it a bit more complicated)

brendan-ward commented on geopandas/io/parquet.py:
```diff
@@ -344,3 +359,97 @@ def _read_parquet(path, columns=None, **kwargs):
+    GeoDataFrame
+    """
+
+    import_optional_dependency(
```
can this now use the 
```
parquet = import_optional_dependency("pyarrow.parquet",...)
```

pattern you used above?

brendan-ward commented on geopandas/io/parquet.py:
```diff
@@ -344,3 +359,97 @@ def _read_parquet(path, columns=None, **kwargs):
+        of the returned GeoDataFrame.  If no geometry columns are present,
+        a ``ValueError`` will be raised.
+    **kwargs
+        Any additional kwargs passed to pyarrow.parquet.read_table().
```
Typo: should be `pyarrow.feather.read_feather`

brendan-ward commented on geopandas/io/parquet.py:
```diff
@@ -344,3 +359,97 @@ def _read_parquet(path, columns=None, **kwargs):
+    """
+
+    feather = import_optional_dependency(
+        "pyarrow.feather", extra="pyarrow is required for Parquet support."
```
Typo, should be ...`Feather support`

jorisvandenbossche commented:
Thanks for the review!

I still need to further update the tests (and properly skip for older pyarrow versions).

jorisvandenbossche commented on geopandas/io/tests/test_arrow.py:
```diff
@@ -357,98 +387,103 @@ def test_parquet_invalid_metadata(tmpdir, geo_meta, error):
         ValueError, match="No geometry columns are included in the columns read"
     ):
-        read_parquet(filename, columns=[])
+        reader(filename, columns=["name"])
```
The Feather reader apparently doesn't work with an empty list for columns selection (it then returns all columns), so changing the test here to only read a non-geometry column

jorisvandenbossche commented:
@brendan-ward this is ready for another look.

I added benchmarks for Feather, but also refactored the IO benchmarks a bit to combine those with Parquet (the setup was basically the same).

Running them now gives:

```
[ 60.00%] ··· io.BenchFrame.time_read                                                                                                                                                                            ok
[ 60.00%] ··· ================ ===========
                file_format               
              ---------------- -----------
               ESRI Shapefile   367±300ms 
                  GeoJSON       619±100ms 
                    GPKG        615±100ms 
                  Parquet        20.8±3ms 
                  Feather        19.5±5ms 
              ================ ===========

[ 70.00%] ··· io.BenchFrame.time_write                                                                                                                                                                           ok
[ 70.00%] ··· ================ ============
                file_format                
              ---------------- ------------
               ESRI Shapefile   1.35±0.08s 
                  GeoJSON       1.33±0.05s 
                    GPKG        1.13±0.2s  
                  Parquet       33.0±10ms  
                  Feather       28.8±20ms  
              ================ ============

[ 80.00%] ··· io.BenchSeries.time_read_series                                                                                                                                                                    ok
[ 80.00%] ··· ================ ==========
                file_format              
              ---------------- ----------
               ESRI Shapefile   463±60ms 
                  GeoJSON       519±40ms 
                    GPKG        512±40ms 
              ================ ==========

[ 90.00%] ··· io.BenchSeries.time_read_series_from_frame                                                                                                                                                         ok
[ 90.00%] ··· ================ ==========
                file_format              
              ---------------- ----------
               ESRI Shapefile   597±80ms 
                  GeoJSON       672±20ms 
                    GPKG        526±30ms 
              ================ ==========

[100.00%] ··· io.BenchSeries.time_write_series                                                                                                                                                                   ok
[100.00%] ··· ================ ==========
                file_format              
              ---------------- ----------
               ESRI Shapefile   837±70ms 
                  GeoJSON       706±60ms 
                    GPKG        754±40ms 
              ================ ==========
```

jorisvandenbossche commented on benchmarks/io.py:
```diff
@@ -1,59 +1,109 @@
 
-    def time_read_frame(self, ext):
-        frame = GeoDataFrame.from_file(self.frame_filename)
+class BenchSeries(Bench):
```
I kept this as a separate class for now (those don't work with Feather/Parquet), but honestly I am not sure it is very useful to keep them .. In the end it is using the same read/write functionality, the only difference is selecting only the geometry column before/after writing/reading.

brendan-ward commented on geopandas/geodataframe.py:
```diff
@@ -600,21 +600,59 @@ def to_parquet(self, filename, compression="snappy", index=None, **kwargs):
             If ``False``, the index(es) will not be written to the file.
             If ``None``, the index(ex) will be included as columns in the file
             output except `RangeIndex` which is stored as metadata only.
+        compression : {'zstd', 'lz4', 'uncompressed'}, optional
```
Apologies for not catching this sooner, but if `compression` is in the docstring it should be in the function parameters too, with the default value:

```
def to_feather(self, path, index=None, compression=None, **kwargs):
```

Also please add default value (and missing `None`) to the docstring, same as for `to_parquet`
```
compression : {'zstd', 'lz4', 'uncompressed', None}, default None.
```

(here and `_to_feather` also)

jorisvandenbossche replied:
> Also please add default value (and missing None) to the docstring, same as for to_parquet

It's a a bit a semantic discussion ;), but the "optional" can also be seen as already mentioning the default of None (in general I think you will see `, default None` and `, optional` quite mixed in projects)

brendan-ward replied:
Sure...  Mostly I was noticing the inconsistency of how this is handled between `to_parquet` and `to_feather` and also `index` parameter immediately above.

jorisvandenbossche commented:
An thanks for the review!
